branch locker test fails: "should not acquire writer while committer is waiting"

"It failed on my machine".  Once.  Uggh.

```
--- FAIL: TestBranchLock (0.94s)
    --- FAIL: TestBranchLock/committer_wait_for_writers (0.00s)
        branch_locker_test.go:158: Should not acquire writer while committer is waiting, err=<nil>
FAIL
coverage: 16.6% of statements
FAIL	github.com/treeverse/lakefs/graveler/ref	6.068s
```

--- a/db/database.go
+++ b/db/database.go
@@ -186,7 +186,8 @@ func (d *PgxDatabase) Transact(ctx context.Context, fn TxFunc, opts ...TxOpt) (i
 		if err != nil {
 			rollbackErr := tx.Rollback(ctx)
 			if rollbackErr != nil {
-				return nil, rollbackErr
+				// returning the original error and not the rollbackErr
+				return nil, err
 			}
 			// retry on serialization error
 			if IsSerializationError(err) {

--- a/graveler/graveler.go
+++ b/graveler/graveler.go
@@ -558,11 +558,6 @@ type StagingManager interface {
 // BranchLockerFunc
 type BranchLockerFunc func() (interface{}, error)
 
-// BranchLocker enforces the branch locking logic
-// The logic is as follows:
-// - Allow concurrent writers to acquire the lock.
-// - A Metadata update waits for all current writers to release the lock, and then gets the lock.
-// - While a metadata update has the lock or is waiting for the lock, any other operation fails to acquire the lock.
 type BranchLocker interface {
 	Writer(ctx context.Context, repositoryID RepositoryID, branchID BranchID, lockedFn BranchLockerFunc) (interface{}, error)
 	MetadataUpdater(ctx context.Context, repositoryID RepositoryID, branchID BranchID, lockeFn BranchLockerFunc) (interface{}, error)

--- a/graveler/ref/branch_locker.go
+++ b/graveler/ref/branch_locker.go
@@ -10,6 +10,8 @@ import (
 	"github.com/treeverse/lakefs/graveler"
 )
 
+// BranchLocker enforces the branch locking logic with Postgres advisory lock
+// The lock can be held by an arbitrary number of Writers or a single MetadataUpdater.
 type BranchLocker struct {
 	db db.Database
 }
@@ -21,40 +23,25 @@ func NewBranchLocker(db db.Database) *BranchLocker {
 }
 
 // Writer tries to acquire a write lock using a Postgres advisory lock for the span of calling `lockedFn`.
-// Returns ErrLockNotAcquired if it cannot acquire the lock or if a commit is in progress.
+// Returns ErrLockNotAcquired if it cannot acquire the lock.
 func (l *BranchLocker) Writer(ctx context.Context, repositoryID graveler.RepositoryID, branchID graveler.BranchID, lockedFn graveler.BranchLockerFunc) (interface{}, error) {
-	writerLockKey, _ := calculateBranchLockerKeys(repositoryID, branchID)
+	writerLockKey := calculateBranchLockerKeys(repositoryID, branchID)
 	return l.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
 		// try to get a shared lock on the writer key
-		var locked bool
-		err := tx.GetPrimitive(&locked, `SELECT pg_try_advisory_xact_lock_shared($1)`, writerLockKey)
+		_, err := tx.Exec(`SELECT pg_advisory_xact_lock_shared($1);`, writerLockKey)
 		if err != nil {
 			return nil, fmt.Errorf("%w (%d): %s", graveler.ErrLockNotAcquired, writerLockKey, err)
-		}
-		if !locked {
-			return nil, fmt.Errorf("%w (%d)", graveler.ErrAlreadyLocked, writerLockKey)
 		}
 		return lockedFn()
 	}, db.WithIsolationLevel(pgx.ReadCommitted))
 }
 
 // MetadataUpdater tries to lock as committer using a Postgres advisory lock for the span of calling `lockedFn`.
-// The call is blocked until all writers end their work.
-// It returns ErrLockNotAcquired if it fails to acquire the lock or if another commit is already in progress.
+// It returns ErrLockNotAcquired if it fails to acquire the lock.
 func (l *BranchLocker) MetadataUpdater(ctx context.Context, repositoryID graveler.RepositoryID, branchID graveler.BranchID, lockedFn graveler.BranchLockerFunc) (interface{}, error) {
-	writerLockKey, committerLockKey := calculateBranchLockerKeys(repositoryID, branchID)
+	writerLockKey := calculateBranchLockerKeys(repositoryID, branchID)
 	return l.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		// try lock committer key
-		var locked bool
-		err := tx.GetPrimitive(&locked, `SELECT pg_try_advisory_xact_lock($1);`, committerLockKey)
-		if err != nil {
-			return nil, fmt.Errorf("%w (%d): %s", graveler.ErrLockNotAcquired, committerLockKey, err)
-		}
-		if !locked {
-			return nil, fmt.Errorf("%w (%d)", graveler.ErrAlreadyLocked, writerLockKey)
-		}
-		// lock writer key
-		_, err = tx.Exec(`SELECT pg_advisory_xact_lock($1);`, writerLockKey)
+		_, err := tx.Exec(`SELECT pg_advisory_xact_lock($1);`, writerLockKey)
 		if err != nil {
 			return nil, fmt.Errorf("%w (%d): %s", graveler.ErrLockNotAcquired, writerLockKey, err)
 		}
@@ -62,13 +49,11 @@ func (l *BranchLocker) MetadataUpdater(ctx context.Context, repositoryID gravele
 	}, db.WithIsolationLevel(pgx.ReadCommitted))
 }
 
-func calculateBranchLockerKeys(repositoryID graveler.RepositoryID, branchID graveler.BranchID) (writerLockKey int64, committerLockKey int64) {
+func calculateBranchLockerKeys(repositoryID graveler.RepositoryID, branchID graveler.BranchID) int64 {
 	h := fnv.New64()
 	_, _ = h.Write([]byte(repositoryID))
 	_, _ = h.Write([]byte{0})
 	_, _ = h.Write([]byte(branchID))
 	_, _ = h.Write([]byte{0})
-	writerLockKey = int64(h.Sum64())
-	committerLockKey = writerLockKey + 1
-	return
+	return int64(h.Sum64())
 }

--- a/graveler/ref/branch_locker.go
+++ b/graveler/ref/branch_locker.go
@@ -25,7 +25,7 @@ func NewBranchLocker(db db.Database) *BranchLocker {
 // Writer tries to acquire a write lock using a Postgres advisory lock for the span of calling `lockedFn`.
 // Returns ErrLockNotAcquired if it cannot acquire the lock.
 func (l *BranchLocker) Writer(ctx context.Context, repositoryID graveler.RepositoryID, branchID graveler.BranchID, lockedFn graveler.BranchLockerFunc) (interface{}, error) {
-	writerLockKey := calculateBranchLockerKeys(repositoryID, branchID)
+	writerLockKey := calculateBranchLockerKey(repositoryID, branchID)
 	return l.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
 		// try to get a shared lock on the writer key
 		_, err := tx.Exec(`SELECT pg_advisory_xact_lock_shared($1);`, writerLockKey)
@@ -39,7 +39,7 @@ func (l *BranchLocker) Writer(ctx context.Context, repositoryID graveler.Reposit
 // MetadataUpdater tries to lock as committer using a Postgres advisory lock for the span of calling `lockedFn`.
 // It returns ErrLockNotAcquired if it fails to acquire the lock.
 func (l *BranchLocker) MetadataUpdater(ctx context.Context, repositoryID graveler.RepositoryID, branchID graveler.BranchID, lockedFn graveler.BranchLockerFunc) (interface{}, error) {
-	writerLockKey := calculateBranchLockerKeys(repositoryID, branchID)
+	writerLockKey := calculateBranchLockerKey(repositoryID, branchID)
 	return l.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
 		_, err := tx.Exec(`SELECT pg_advisory_xact_lock($1);`, writerLockKey)
 		if err != nil {
@@ -49,7 +49,7 @@ func (l *BranchLocker) MetadataUpdater(ctx context.Context, repositoryID gravele
 	}, db.WithIsolationLevel(pgx.ReadCommitted))
 }
 
-func calculateBranchLockerKeys(repositoryID graveler.RepositoryID, branchID graveler.BranchID) int64 {
+func calculateBranchLockerKey(repositoryID graveler.RepositoryID, branchID graveler.BranchID) int64 {
 	h := fnv.New64()
 	_, _ = h.Write([]byte(repositoryID))
 	_, _ = h.Write([]byte{0})

--- a/graveler/ref/branch_locker_test.go
+++ b/graveler/ref/branch_locker_test.go
@@ -7,12 +7,15 @@ import (
 	"sync"
 	"sync/atomic"
 	"testing"
+	"time"
 
 	"github.com/treeverse/lakefs/graveler"
 	"github.com/treeverse/lakefs/graveler/ref"
 	"github.com/treeverse/lakefs/graveler/testutil"
 	tu "github.com/treeverse/lakefs/testutil"
 )
+
+var errUnexpectedCall = errors.New("this function should not have been called")
 
 func TestBranchLock(t *testing.T) {
 	conn, _ := tu.GetDB(t, databaseURI)
@@ -48,13 +51,14 @@ func TestBranchLock(t *testing.T) {
 		}
 	})
 
-	t.Run("committer_blocks_all", func(t *testing.T) {
-		chAcquired := make(chan struct{})
+	t.Run("committer_blocks_writer", func(t *testing.T) {
 		chReleaseAcquired := make(chan struct{})
-		chDone := make(chan struct{})
+		chAcquired := make(chan struct{})
+		defer close(chReleaseAcquired)
+		// call writer and wait on channel
+		ctx := context.Background()
 		go func() {
-			ctx := context.Background()
-			_, err := bl.MetadataUpdater(ctx, "b", testutil.DefaultBranchID, func() (interface{}, error) {
+			_, err := bl.MetadataUpdater(ctx, "committer_blocks_writer", testutil.DefaultBranchID, func() (interface{}, error) {
 				close(chAcquired)
 				<-chReleaseAcquired
 				return nil, nil
@@ -62,120 +66,46 @@ func TestBranchLock(t *testing.T) {
 			if err != nil {
 				t.Error("Metadata updater request failed:", err)
 			}
-			close(chDone)
 		}()
-
-		ctx := context.Background()
-		// wait until we acquire metadata update lock
 		<-chAcquired
-		// try to acquire writer
-		_, err := bl.Writer(ctx, "b", testutil.DefaultBranchID, func() (interface{}, error) {
-			return nil, nil
+		// check Writer waits (context gets to deadline before Writer callback is called)
+		timeToDeadline := time.Now().Add(time.Second)
+		ctxWithDeadline, _ := context.WithDeadline(ctx, timeToDeadline)
+		_, err := bl.Writer(ctxWithDeadline, "committer_blocks_writer", testutil.DefaultBranchID, func() (interface{}, error) {
+			return nil, errUnexpectedCall
 		})
-		if !errors.Is(err, graveler.ErrAlreadyLocked) {
-			t.Fatalf("Writer should be locked during metadata updater, err=%v", err)
-		}
-		// try to acquire committer
-		_, err = bl.MetadataUpdater(ctx, "b", testutil.DefaultBranchID, func() (interface{}, error) {
-			return nil, nil
-		})
-		if !errors.Is(err, graveler.ErrAlreadyLocked) {
-			t.Fatalf("Committer should be locked during metadata updater, err=%s", err)
-		}
-		// release acquired lock
-		close(chReleaseAcquired)
-		<-chDone
-		// check we can write
-		_, err = bl.Writer(ctx, "b", testutil.DefaultBranchID, func() (interface{}, error) {
-			return nil, nil
-		})
-		if err != nil {
-			t.Fatalf("Failed to acquire writer, err=%s", err)
+		if !errors.Is(err, graveler.ErrLockNotAcquired) {
+			t.Errorf("unexpected error got: %v expected: %s", err, graveler.ErrLockNotAcquired)
 		}
 	})
 
-	t.Run("committer_wait_for_writers", func(t *testing.T) {
-		// start a writer and block it
-		chAcquireWriter := make(chan struct{})
-		chReleaseWriter := make(chan struct{})
-		chDoneWriter := make(chan struct{})
+	t.Run("writer_blocks_commiter", func(t *testing.T) {
+		chReleaseAcquired := make(chan struct{})
+		defer close(chReleaseAcquired)
+		chAcquired := make(chan struct{})
+		// call writer and wait on channel
+		ctx := context.Background()
 		go func() {
-			ctx := context.Background()
-			_, err := bl.Writer(ctx, "c", testutil.DefaultBranchID, func() (interface{}, error) {
-				close(chAcquireWriter)
-				<-chReleaseWriter
+			_, err := bl.Writer(ctx, "writer_blocks_commiter", testutil.DefaultBranchID, func() (interface{}, error) {
+				close(chAcquired)
+				<-chReleaseAcquired
 				return nil, nil
 			})
 			if err != nil {
-				t.Errorf("Failed to acquire writer: %s", err)
+				t.Error("Metadata updater request failed:", err)
 			}
-			close(chDoneWriter)
 		}()
-		<-chAcquireWriter
+		<-chAcquired
+		// check MetadataUpdater waits ( context gets to deadline before MetadataUpdater callback is called)
+		timeToDeadline := time.Now().Add(time.Second)
+		ctxWithDeadline, cancel := context.WithDeadline(ctx, timeToDeadline)
+		defer cancel()
 
-		// start two committers and wait until one of them will fail to know that the first one is blocked
-		var metadataUpdates int64
-		chDoneCommitter := make([]chan struct{}, 2)
-		chReleaseCommitter := make(chan struct{})
-		committersErr := make([]error, 2)
-		for i := 0; i < len(chDoneCommitter); i++ {
-			chDoneCommitter[i] = make(chan struct{})
-			go func(pos int) {
-				ctx := context.Background()
-				_, committersErr[pos] = bl.MetadataUpdater(ctx, "c", testutil.DefaultBranchID, func() (interface{}, error) {
-					atomic.AddInt64(&metadataUpdates, 1)
-					<-chReleaseCommitter
-					return nil, nil
-				})
-				close(chDoneCommitter[pos])
-			}(i)
-		}
-
-		// wait for one of the committers
-		var failedCommitterPos int
-		select {
-		case <-chDoneCommitter[0]:
-			failedCommitterPos = 0
-		case <-chDoneCommitter[1]:
-			failedCommitterPos = 1
-		}
-
-		// check that the one that failed - failed with the right reason
-		if err := committersErr[failedCommitterPos]; !errors.Is(err, graveler.ErrAlreadyLocked) {
-			t.Errorf("Failed committer (%d) should failed to acquire, err=%v", failedCommitterPos, err)
-		}
-		updates := atomic.LoadInt64(&metadataUpdates)
-		if updates != 0 {
-			t.Errorf("No update should be done at this point, updates=%d", updates)
-		}
-
-		// verify that another writer can't start while committer is waiting
-		ctx := context.Background()
-		_, err := bl.Writer(ctx, "c", testutil.DefaultBranchID, func() (interface{}, error) {
-			return nil, nil
+		_, err := bl.MetadataUpdater(ctxWithDeadline, "writer_blocks_commiter", testutil.DefaultBranchID, func() (interface{}, error) {
+			return nil, errUnexpectedCall
 		})
-		if !errors.Is(err, graveler.ErrAlreadyLocked) {
-			t.Errorf("Should not acquire writer while committer is waiting, err=%v", err)
-		}
-
-		// release the last writer and wait for it
-		close(chReleaseWriter)
-		<-chDoneWriter
-
-		// release that running committer
-		close(chReleaseCommitter)
-
-		// wait for the second committer
-		secondCommitterPos := (failedCommitterPos + 1) % 2
-		<-chDoneCommitter[secondCommitterPos]
-
-		// verify no error and one update
-		if err := committersErr[secondCommitterPos]; err != nil {
-			t.Errorf("Committer should ended without an error, err=%v", err)
-		}
-		updates = atomic.LoadInt64(&metadataUpdates)
-		if updates != 1 {
-			t.Errorf("Expected one update, updates=%d", updates)
+		if !errors.Is(err, graveler.ErrLockNotAcquired) {
+			t.Errorf("unexpected error got: %v expected: %s", err, graveler.ErrLockNotAcquired)
 		}
 	})
 }

